apps/src/dance/ai/DanceAiModal.tsx

```diff
@@ -440,11 +496,63 @@ const DanceAiModal: React.FunctionComponent = () => {
       return <span key={index}>{part}</span>;
     });
 
-  const currentGeneratedEffect = getCurrentGeneratedEffect();
+  const currentGeneratedEffect = getGeneratedEffect(generatingProgress.step);
 
   const lastInputItem =
     currentInputSlot > 0 ? getItem(inputs[currentInputSlot - 1]) : undefined;
 
+  // Visualization preview size, in pixels.
+  const previewSize = 280;
+  const previewSizeSmall = 90;
+
+  const labels = getLabels();
+
+  // Calculates the minimum individual score
+  // (ie, a SINGLE emoji association with a foreground/background palette combination),
+  // and a maximum total score
+  // (ie, the sum of ALL selected emoji's associations with a foreground/background palette combination).
+  // Used to normalize and scale the data for easier differentiation between results by the user.
+  const calculateMinMax = () => {
+    // The minimum individual score is selected across all generated effects (bad and good).
+    const minIndividualScore = Array.from(
+      Array(BAD_GENERATED_RESULTS_COUNT + 1).keys()
+    ).reduce((accumulator: number, currentValue: number) => {
+      const scores = getScores(currentValue);
+      const min = Math.min(...scores);
+      return min < accumulator ? min : accumulator;
+    }, Infinity);
+
+    // By definition, the maximum total score must come from the "good" effect.
+    const goodEffectScores = getScores(BAD_GENERATED_RESULTS_COUNT);
+    const maxTotalScore = Math.max(
+      goodEffectScores.reduce((sum, score) => (sum += score))
+    );
+
+    return {minIndividualScore, maxTotalScore};
+  };
+
+  if (
+    generatedEffects.current.goodEffect &&
+    generatedEffects.current.badEffects.length === BAD_GENERATED_RESULTS_COUNT
+  ) {
+    minMaxAssociations.current = calculateMinMax();
```

Comment:
`useState` does seem to favor compound states because of the way it re-renders even when only one is changed, but I must admit it feels inconvenient compared to the old way React handled the state object.  Do we have a good example of this?

Reply:
I believe there's one min/max association across all generating steps.

Reply:
@breville yeah, not sure if there's a good example anywhere, but as another option we could look into using [useReducer](https://react.dev/reference/react/useReducer) which allows for redux-like reducer logic in a hook. Of course another option would be just to create a redux store for this if we feel it would be useful.